packages/services/tokens/src/multi-tier-storage.ts

```diff
@@ -0,0 +1,395 @@
+      const context = { maskedToken, source: 'in-memory' };
+      const data = await cache.fetch(hashedToken, {
+        context,
+        status,
+      });
+
+      if (status.fetch) {
+        recordCacheRead(status.fetch);
+      } else {
+        serverLogger.warn('Status of the fetch is missing');
+      }
+
+      if (!data) {
+        // Looked up in all layers, and the token is not found
+        return null;
+      }
+
+      if (data === 'not-found') {
+        return null;
+      }
+
+      touch.schedule(hashedToken);
+      return data;
+    },
+    writeToken: tracker.wrap(async item => {
+      const result = await db.createToken(item);
+
+      // We don't want to write to in-memory cache,
+      // as the token may not be used immediately, or at all.
+      // We write to Redis, so in case the token is used,
+      // we reuse it for the in-memory cache, without hitting the DB.
+      const hashedToken = result.token;
+      const cacheEntry = transformToken(result);
+
+      // Write to Redis gracefully. If it fails, we log the error, but we don't throw.
+      // The token won't be in Redis cache, but it will be possible to retrieve it from the DB.
+      // It will affect performance slightly, but it won't break the system.
+      try {
+        await setInRedis(redis, hashedToken, cacheEntry);
+      } catch (error) {
+        handleStorageError({
+          logger: serverLogger,
+          error,
+          logMsg: 'Failed to write token to Redis, but it was created in the Db',
+          tier: 'redis',
+          action: 'set',
+        });
+      }
+
+      return cacheEntry;
+    }),
+    deleteToken: tracker.wrap(async hashedToken => {
+      await db.deleteToken({
+        token: hashedToken,
+        async postDeletionTransaction() {
+          // delete from Redis when the token is deleted from the DB
+          await redis.del([generateRedisKey(hashedToken), generateStaleRedisKey(hashedToken)]);
+          // only then delete from the in-memory cache.
+          // The other replicas will purge the token from
+          // their own in-memory caches on their own pace (ttl)
+          cache.delete(hashedToken);
+        },
+      });
+    }),
+    invalidateTokens: tracker.wrap(async hashedTokens => {
+      if (hashedTokens.length === 0) {
+        return;
+      }
+
+      await redis.del(
+        hashedTokens.map(generateRedisKey).concat(hashedTokens.map(generateStaleRedisKey)),
+      );
+      for (const hashedToken of hashedTokens) {
+        cache.delete(hashedToken);
+      }
+    }),
+  };
+}
+
+function transformToken(item: tokens): StorageItem {
+  return {
+    token: item.token,
+    tokenAlias: item.token_alias,
+    name: item.name,
+    date: item.created_at as any,
+    lastUsedAt: item.last_used_at as any,
+    organization: item.organization_id,
+    project: item.project_id,
+    target: item.target_id,
+    scopes: item.scopes || [],
+  };
+}
```

Comment:
just a nit, but instead of transforming the tokens after reading from db, we should probably do it as part of the db logic already 🤔

Reply:
done